x-pack/plugins/files/server/blob_storage_service/adapters/es/content_stream/content_stream.ts

```diff
@@ -0,0 +1,397 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import cuid from 'cuid';
+import * as cborx from 'cbor-x';
+import { errors as esErrors } from '@elastic/elasticsearch';
+import type { ElasticsearchClient, Logger } from '@kbn/core/server';
+import { ByteSizeValue } from '@kbn/config-schema';
+import { defaults } from 'lodash';
+import { Duplex, Writable, Readable } from 'stream';
+
+import type { FileChunkDocument } from '../mappings';
+
+/**
+ * @note The Elasticsearch `http.max_content_length` is including the whole POST body.
+ * But the update/index request also contains JSON-serialized query parameters.
+ * 1Kb span should be enough for that.
+ */
+const REQUEST_SPAN_SIZE_IN_BYTES = 1024;
+
+type Callback = (error?: Error) => void;
+
+export type ContentStreamEncoding = 'base64' | 'raw';
+
+interface IndexRequestParams {
+  data: Buffer;
+  id: string;
+  index: string;
+  bid: string;
+}
+
+export interface ContentStreamParameters {
+  /**
+   * The maximum size allowed per chunk.
+   *
+   * @default 4mb
+   */
+  maxChunkSize?: string;
+  /**
+   * The file size in bytes. This can be used to optimize downloading.
+   */
+  size?: number;
+}
+
+export class ContentStream extends Duplex {
+  /**
+   * @see https://en.wikipedia.org/wiki/Base64#Output_padding
+   */
+  private static getMaxBase64EncodedSize(max: number) {
+    return Math.floor(max / 4) * 3;
+  }
+
+  private buffers: Buffer[] = [];
+  private bytesBuffered = 0;
+
+  private bytesRead = 0;
+  private chunksRead = 0;
+  private chunksWritten = 0;
+  private maxChunkSize?: number;
+  private parameters: Required<ContentStreamParameters>;
+
+  /**
+   * The number of bytes written so far.
+   * Does not include data that is still queued for writing.
+   */
+  bytesWritten = 0;
+
+  constructor(
+    private readonly client: ElasticsearchClient,
+    private id: undefined | string,
+    private readonly index: string,
+    private readonly logger: Logger,
+    parameters: ContentStreamParameters = {}
+  ) {
+    super();
+    this.parameters = defaults(parameters, {
+      encoding: 'base64',
+      size: -1,
+      maxChunkSize: '4mb',
+    });
+  }
+
+  private getMaxContentSize(): number {
+    return ByteSizeValue.parse(this.parameters.maxChunkSize).getValueInBytes();
+  }
+
+  private getMaxChunkSize() {
+    if (!this.maxChunkSize) {
+      const maxContentSize = this.getMaxContentSize() - REQUEST_SPAN_SIZE_IN_BYTES;
+      this.maxChunkSize = ContentStream.getMaxBase64EncodedSize(maxContentSize);
+      this.logger.debug(`Chunk size is ${this.maxChunkSize} bytes.`);
+    }
```

Comment:
I guess the string-to-number byte parsing is legacy from the Reporting plugin. I'm curious, can we simplify this all such that chunkSize is always bytes and always a number?

```suggestion
      maxChunkSize: 4 * 1024 * 1024,
    });
  }
```